summingbird-storm/src/main/scala/com/twitter/summingbird/storm/SpoutProvider.scala

```diff
@@ -0,0 +1,158 @@
+import com.twitter.summingbird.online.{ Externalizer, executor }
+import com.twitter.summingbird.online.option.{ SourceParallelism, SummerBuilder }
+import com.twitter.summingbird.option.JobId
+import com.twitter.summingbird.planner.{ Dag, SummerNode }
+import com.twitter.summingbird.storm.planner.StormNode
+import com.twitter.tormenta.spout.Spout
+import com.twitter.summingbird.storm.Constants
+import com.twitter.summingbird.storm.option.SpoutStormMetrics
+import com.twitter.summingbird.storm.spout.KeyValueSpout
+
+case class SpoutProvider(storm: Storm, stormDag: Dag[Storm], node: StormNode, jobID: JobId)(implicit topologyBuilder: TopologyBuilder) {
+
+  private def getOrElse[T <: AnyRef: Manifest](default: T): T = storm.getOrElse(stormDag, node, default)
+
+  private def getSourceParallelism(sourceParOption: Option[SourceParallelism]) =
+    getOrElse(sourceParOption.getOrElse(Constants.DEFAULT_SOURCE_PARALLELISM)).parHint
+
+  private val spoutStormMetrics = getOrElse(Constants.DEFAULT_SPOUT_STORM_METRICS)
+
+  private def extractSourceAttributes: (Spout[(Timestamp, Any)], Option[SourceParallelism]) =
+    node.members.collect { case Source(SpoutSource(s, parOpt)) => (s, parOpt) }.head
+
+  private def computeSpout(spout: Spout[(Timestamp, Any)]): Spout[(Timestamp, Any)] = {
+    val nodeName = stormDag.getNodeName(node)
+    node.members.reverse.foldLeft(spout.asInstanceOf[Spout[(Timestamp, Any)]]) { (spout, p) =>
+      p match {
+        case Source(_) => spout // The source is still in the members list so drop it
+        case OptionMappedProducer(_, op) =>
+          val boxed = Externalizer(op)
+          spout.flatMap { case (time, t) => boxed.get(t).map { x => (time, x) } }
+        case FlatMappedProducer(_, op) => {
+          val lockedOp = Externalizer(op)
+          spout.flatMap { case (time, t) => lockedOp.get.apply(t).map { x => (time, x) } }
+        }
+        case NamedProducer(_, _) => spout
+        case IdentityKeyedProducer(_) => spout
+        case AlsoProducer(_, _) => spout
+        case _ => sys.error("not possible, given the above call to span.\n" + p)
+      }
+    }
+  }
+
+  /**
+   * This method checks that sourceParallelism is greater than flatMapParallelism when FMMergeableWithSource is opted-in
+   */
+  private def validateParallelisms(sourceParallelism: Int): Unit = {
+    val isMergeableWithSource = getOrElse(Constants.DEFAULT_FM_MERGEABLE_WITH_SOURCE).get
+    if (isMergeableWithSource) {
+      val flatMapParallelism = getOrElse(Constants.DEFAULT_FM_PARALLELISM).parHint
+      require(flatMapParallelism <= sourceParallelism,
+        s"SourceParallelism ($sourceParallelism) must be at least as high as FlatMapParallelism ($flatMapParallelism) when merging flatMap with Source")
+    }
+  }
+
+  /*
+   * This method formats the tuples to batched format and wraps the spout into KeyValueSpout.
+   */
+  private def createSpoutToFeedSummer[K, V](
+    node: StormNode,
```

Comment:
This shouldn't be called `node` since you already have a `node` in scope.  Actually this already caused me to have you introduce a subtle bug by removing the `StormNode` param in `getOrElse`.

I would:

1. change this to `summerNode`
1. change this to type `SummerNode[Storm]`
1. modify `getOrElse` to take a `StormNode` again